pyo3-macros-backend/src/intopyobject.rs

```diff
@@ -0,0 +1,466 @@
+use crate::attributes::{self, get_pyo3_options, CrateAttribute};
+use crate::utils::Ctx;
+use proc_macro2::{Span, TokenStream};
+use quote::{format_ident, quote, quote_spanned};
+use syn::ext::IdentExt;
+use syn::parse::{Parse, ParseStream};
+use syn::spanned::Spanned as _;
+use syn::{parse_quote, Attribute, DataEnum, DeriveInput, Fields, Ident, Index, Result, Token};
+
+/// Attributes for deriving `IntoPyObject` scoped on containers.
+enum ContainerPyO3Attribute {
+    /// Treat the Container as a Wrapper, directly convert its field into the output object.
+    Transparent(attributes::kw::transparent),
+    /// Change the path for the pyo3 crate
+    Crate(CrateAttribute),
+}
+
+impl Parse for ContainerPyO3Attribute {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(attributes::kw::transparent) {
+            let kw: attributes::kw::transparent = input.parse()?;
+            Ok(ContainerPyO3Attribute::Transparent(kw))
+        } else if lookahead.peek(Token![crate]) {
+            input.parse().map(ContainerPyO3Attribute::Crate)
+        } else {
+            Err(lookahead.error())
+        }
+    }
+}
+
+#[derive(Default)]
+struct ContainerOptions {
+    /// Treat the Container as a Wrapper, directly convert its field into the output object.
+    transparent: Option<attributes::kw::transparent>,
+    /// Change the path for the pyo3 crate
+    krate: Option<CrateAttribute>,
+}
+
+impl ContainerOptions {
+    fn from_attrs(attrs: &[Attribute]) -> Result<Self> {
+        let mut options = ContainerOptions::default();
+
+        for attr in attrs {
+            if let Some(pyo3_attrs) = get_pyo3_options(attr)? {
+                pyo3_attrs
+                    .into_iter()
+                    .try_for_each(|opt| options.set_option(opt))?;
+            }
+        }
+        Ok(options)
+    }
+
+    fn set_option(&mut self, option: ContainerPyO3Attribute) -> syn::Result<()> {
+        macro_rules! set_option {
+            ($key:ident) => {
+                {
+                    ensure_spanned!(
+                        self.$key.is_none(),
+                        $key.span() => concat!("`", stringify!($key), "` may only be specified once")
+                    );
+                    self.$key = Some($key);
+                }
+            };
+        }
+
+        match option {
+            ContainerPyO3Attribute::Transparent(transparent) => set_option!(transparent),
+            ContainerPyO3Attribute::Crate(krate) => set_option!(krate),
+        }
+        Ok(())
+    }
+}
+
+struct IntoPyObjectImpl {
+    target: TokenStream,
+    output: TokenStream,
+    error: TokenStream,
+    body: TokenStream,
+}
+
+struct NamedStructField<'a> {
```

Comment:
We should consider the interaction with `FromPyObject`'s options like `#[pyo3(item)]` and `#[pyo3(attribute)]`. I am fine if we just parse them and reject cases where we don't want to support them for `IntoPyObject (yet)`; all I'd like to avoid is that we don't silently ignore those options.

If there is an unsupported case then users can split their `FromPyObject` and `IntoPyObject` derives onto different structs, and they are then aware their conversions will not round-trip.

Reply:
Hmm, we could definitely parse them, but I'm not sure what we would do with them in this case. Due to the `PyDict` return type this always uses `set_item`. So I think it makes only sense to parse them, if there is a way (I'm not aware of any) to create an anonymous object and set arbitrary attributes on it. I guess we could accept `item` to allow renaming and to make a round-trip possible _if_ it is given.

Reply:
Good point. Let's handle `item` round trips here and error on `attribute`? Users can always make two separate types if they need `#[attribute]` for their `FromPyObject` implementation for some reason.

Reply:
I added parsing for `#[pyo3(item)]` and `#[pyo3(attribute)]` (like for `FromPyObject`), roundtrip tests in `test_frompy_intopy_roundtrip.rs` and ui tests for the unsupported cases.